core/src/Streamly/Internal/Data/Serialize.hs

```diff
@@ -0,0 +1,99 @@
+module Streamly.Internal.Data.Serialize
+    ( Size(..)
+    , Serialize(..)
+    ) where
+
+--------------------------------------------------------------------------------
+-- Imports
+--------------------------------------------------------------------------------
+
+import Control.Monad (void)
+import Data.List (foldl')
+import Data.Proxy (Proxy (..))
+import Streamly.Internal.Data.Unboxed (MutableByteArray(..))
+
+import GHC.Int (Int16(..), Int32(..), Int64(..), Int8(..))
+import GHC.Word (Word16(..), Word32(..), Word64(..), Word8(..))
+import GHC.Stable (StablePtr(..))
+
+import qualified Streamly.Internal.Data.Unboxed as Unbox
+
+import GHC.Exts
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+data Size a
+    = VarSize (a -> Int)
+    | ConstSize (Proxy a -> Int)
+
+class Serialize a where
+    size :: Size a
+    peekByteIndex :: Int -> MutableByteArray -> IO (Int, a)
+    pokeByteIndex :: Int -> MutableByteArray -> a -> IO Int
+
+--------------------------------------------------------------------------------
+-- Instances
+--------------------------------------------------------------------------------
+
+#define DERIVE_SERIALIZE_FROM_UNBOX(_type) \
+instance Serialize _type where \
+;    size = ConstSize Unbox.sizeOf \
+;    peekByteIndex off arr = \
+        Unbox.peekByteIndex off arr >>= \
+            \val -> let sz = Unbox.sizeOf (Proxy :: Proxy _type) \
+                     in pure (off + sz, val) \
+;    pokeByteIndex off arr val = \
+        Unbox.pokeByteIndex off arr val \
+            >> let sz = Unbox.sizeOf (Proxy :: Proxy _type) \
+                in pure (off + sz)
+
+DERIVE_SERIALIZE_FROM_UNBOX(Char)
+DERIVE_SERIALIZE_FROM_UNBOX(Int8)
+DERIVE_SERIALIZE_FROM_UNBOX(Int16)
+DERIVE_SERIALIZE_FROM_UNBOX(Int32)
+DERIVE_SERIALIZE_FROM_UNBOX(Int)
+DERIVE_SERIALIZE_FROM_UNBOX(Int64)
+DERIVE_SERIALIZE_FROM_UNBOX(Word)
+DERIVE_SERIALIZE_FROM_UNBOX(Word8)
+DERIVE_SERIALIZE_FROM_UNBOX(Word16)
+DERIVE_SERIALIZE_FROM_UNBOX(Word32)
+DERIVE_SERIALIZE_FROM_UNBOX(Word64)
+DERIVE_SERIALIZE_FROM_UNBOX(Double)
+DERIVE_SERIALIZE_FROM_UNBOX(Float)
+DERIVE_SERIALIZE_FROM_UNBOX((StablePtr a))
+DERIVE_SERIALIZE_FROM_UNBOX((Ptr a))
+DERIVE_SERIALIZE_FROM_UNBOX((FunPtr a))
+
+instance forall a. Serialize a => Serialize [a] where
+
+    size = VarSize $ \lst ->
+        case size :: Size a of
+            VarSize f ->
+                foldl'
+                    (\acc x -> acc + f x)
+                    (Unbox.sizeOf (Proxy :: Proxy Int))
+                    lst
+            ConstSize f ->
+                length lst
+                    * f (Proxy :: Proxy a)
+                    + Unbox.sizeOf (Proxy :: Proxy Int)
+
+    peekByteIndex off arr = do
+        len <- Unbox.peekByteIndex off arr :: IO Int
+        let off1 = off + Unbox.sizeOf (Proxy :: Proxy Int)
+        let peekList buf o 0 = pure (o, reverse buf)
```

Comment:
In future, for the cases when the order does not matter we can define an alternative instance that does not reverse. That can be done on a newtype wrapper over list.